src/core/execute.c

```diff
@@ -268,25 +268,26 @@ static int exec_context_tty_size(const ExecContext *context, unsigned *ret_rows,
 }
 
 static void exec_context_tty_reset(const ExecContext *context, const ExecParameters *p) {
-        const char *path;
+        _cleanup_close_ int fd = -EBADF;
+        const char *path = exec_context_tty_path(ASSERT_PTR(context));
 
-        assert(context);
+        /* Take a lock around the device for the duration of the setup that we do here.
+         * systemd-vconsole-setup.service also takes the lock to avoid being interrupted.
+         * We open a new fd that will be closed automatically, and operate on it for convenience.
+         */
 
-        path = exec_context_tty_path(context);
+        if (p && p->stdin_fd >= 0)
+                fd = xopenat_lock(fd, NULL,
+                                  O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY, 0, 0, LOCK_BSD, LOCK_EX);
+        else if (path)
+                fd = xopenat_lock(AT_FDCWD, path,
+                                  O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY, 0, 0, LOCK_BSD, LOCK_EX);
```

Comment:
This drops the retry logic from `open_terminal()`, is that intended?

Maybe add `open_terminal_lock()` that works in `openat()` style that takes fd, path, lock type and lock operation and calls xopenat_lock() internally. And `open_terminal()` can just call `open_terminal_lock()` with `LOCK_NONE` then.

Then you can have both these calls call `open_terminal_lock()` to either reopen the fd or open the path and lock it

Reply:
I don't think it makes sense to create a helper with a single user… I reverted that chunk of code to the previous version. I also changed the code to bail out if we cannot open or lock the fd/path. The operations are optional anyway, so I think it's better to just skip if the initial steps fail.